blocks/serialization.py

```diff
@@ -1,65 +1,167 @@
+from theano.misc import pkl_utils
+from theano.misc.pkl_utils import (PersistentCudaNdarrayID,
+                                   PersistentSharedVariableID)
+
+from blocks.config import config
+from blocks.filter import get_brick
+from blocks.utils import change_recursion_limit
+
 
-from blocks.utils import reraise_as
+BRICK_DELIMITER = '-'
+MAIN_MODULE_WARNING = """WARNING: Main loop depends on the function `{}` in \
+`__main__` namespace.
 
-PICKLING_ERROR = """
+Because of limitations to pickling, this means that you will not be able to \
+resume your model outside of a namespace containing this function. In other \
+words, you can only call `continue_training` from within this script."""
 
-Blocks relies on the ability to pickle the entire main loop, which includes \
-all bricks, data streams, extensions, etc. All of these must be serializable \
-using Pythons pickling library. This means certain things such as nested \
-functions, lambda expressions, generators, etc. should be avoided. Please \
-see the documentation for more detail."""
 
-INSTANCEMETHOD_ERROR = """
+class PersistentParameterID(PersistentSharedVariableID):
+    """Persist the names of parameter arrays in the zip file.
 
-Python is unable to serialize references to class and instance methods. This \
-means that you cannot store reference to objects like `str.lower` (a class \
-method) or `Linear().allocate` (an instance method) in an attribute. If you \
-reall need to do this, it is often possible to write a wrapper function such \
-as `def lower(s): return s.lower()` and store that instead."""
+    Only Theano shared variables are persisted to the zip file using this
+    method. Names are determined using the brick hierarchy, or the shared
+    variable name.
 
-LAMBDA_ERROR = """
+    Parameters
+    ----------
+    allow_unnamed : bool, optional
+        Allow shared variables without a name to be persisted. Defaults to
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-Python is unable to serialize lambda functions. Make sure that your code does \
-not rely on them and use normal function definitions instead."""
+    """
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            super(PersistentParameterID, self).__call__(obj)
+            if hasattr(obj.tag, 'annotations'):
+                name = '{}.{}'.format(
+                    BRICK_DELIMITER.join([brick.name for brick in
+                                          get_brick(obj).get_unique_path()]),
+                    [obj.name]
```

Comment:
@dmitriy-serdyuk The current error is my bad, this should be `obj.name` instead of `[obj.name]`